docs/tutorial/Zmod37.lean

```diff
@@ -0,0 +1,195 @@
+/- Integers mod 37
+
+  A demonstration of how to use equivalence relations and equivalence classes in Lean.
+
+  We define the "congruent mod 37" relation on integers, prove it is an equivalence
+  relation, define Zmod37 to be the equivalence classes, and put a ring structure on
+  the quotient.
+
+-/
+-- this import is helpful for some intermediate calculation
+import tactic.ring
+
+-- Definition of the equivalence relation
+definition cong_mod37 (a b : ℤ) : Prop := ∃ (k : ℤ), k * 37 = b - a
+
+-- Now check it's an equivalence reln!
+
+theorem cong_mod_refl : reflexive (cong_mod37) :=
+begin
+  intro x,
+  -- to prove cong_mod37 x x we just observe that k = 0 will do.
+  use (0:ℤ), -- this is k
+  simp,
+end
+
+theorem cong_mod_symm : symmetric (cong_mod37) :=
+begin
+  intros a b H,
+  -- H : cond_mod37 a b
+  cases H with k Hk,
+  -- Hk : k * 37 = (b - a)
+  -- Goal is to find an integer k' with k' * 37 = a - b  
+  use -k,
+  simp [Hk],
+end
+
+theorem cong_mod_trans : transitive (cong_mod37) :=
+begin
+  intros a b c Hab Hbc,
+  cases Hab with k Hk,
+  cases Hbc with l Hl,
+  -- Hk : k*37 = b - a, and Hl : l*37 = c - b
+  -- Goal : m * 37 = c - a
+  use (k+l),
+  simp [add_mul,Hk,Hl],
+end
+
+-- so we've now seen a general technique for proving a ≈ b -- use (the k that works)
+
+theorem cong_mod_equiv : equivalence (cong_mod37) :=
+⟨cong_mod_refl,cong_mod_symm,cong_mod_trans⟩
+
+instance Z_setoid : setoid ℤ := { r := cong_mod37, iseqv := cong_mod_equiv }
```

Comment:
Could you add a comment that putting a global setoid instance on Z is OK for a tutorial, but definitely a bad idea in the wild, and that one would use `def` and `local attribute [instance]` in a section instead?

Reply:
Chris pointed out that `instance coe_int_Zmod37 : has_coe ℤ (Zmod37) := ⟨reduce_mod37⟩` is an issue in the wild, because once Zmod37 is a ring there will be a coercion from int anyway. I'll deal with these things soon.

Reply:
I have made some troublesome instances local.